### PR TITLE
ebpf_updates: issue 4

### PR DESCRIPTION
**Pull request by qmonnet**
Fourth issue of the eBPF Updates, for ~February 2021

### CODE REVIEW

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+  Various aspects of eBPF are covered in this presentation. After introducing
+  the basics, Datadog discusses technical details related to their workflow: Is
+  it better to guess the offset of kernel structures or to rely on runtime
+  compilation? They also describe two of their use cases for eBPF, network
+  monitoring and runtime security.
+
+<!--2021-02-09-->
+* [_A Beginner's Guide to eBPF with Go_](https://web.archive.org/web/20210216233137/https://files.gotocon.com/uploads/slides/conference_39/1688/original/Beginners%20guide%20to%20eBPF%20with%20Go.pdf) (PDF),
+  from Liz Rice.  
+  Learn how to program with eBPF and Golang with this accessible tutorial.
```
Comment (tklauser) [inline src/posts/2021-02-22-ebpf-updates/index.md:146]:
```suggestion
  Learn how to program with eBPF and Go with this accessible tutorial.
```

c.f. https://golang.org/doc/faq#go_or_golang

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+  details are available in
+  [the editor's announcement](https://cilium.io/blog/2021/02/10/network-policy-editor).
+
+* [Tracee version 0.5.0](https://github.com/aquasecurity/tracee/discussions/563)
+  is out. From a tracing command line tool, the project evolves into a runtime
+  security solution and now encompasses tracee-ebpf—the command-line tool
+  itself—and the libbpfgo library. It also includes tracee-rules, a “new rule
+  engine to process tracee-ebpf's events and detects suspicious behavior based
+  on built-in and user-defined 'signatures'”, which are defined in Open Policy
+  Agent's Rego language or in Golang.
```
Comment (tklauser) [inline src/posts/2021-02-22-ebpf-updates/index.md:180]:
```suggestion
  Agent's Rego language or in Go.
```

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+
+* Guardicore released [IPCDump](https://github.com/guardicore/IPCDump) for
+  tracing interprocess communications on Linux, be it through pipes, FIFOs,
+  signals, UNIX sockets, loopback-based networking, or pseudoterminals. It
+  draws some inspiration from Windows's procmon. Internally, it uses the gobpf
+  library that provides Golang bindings for the BCC framework. See also
```
Comment (tklauser) [inline src/posts/2021-02-22-ebpf-updates/index.md:186]:
```suggestion
  library that provides Go bindings for the BCC framework. See also
```

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+  is a note from Peter Ruderich describing a small but standalone XDP example,
+  with a few explanations and pointers. This is an interesting program to get
+  started with XDP (but don't forget the
+  [XDP tutorial](https://github.com/xdp-project/xdp-tutorial/)).
+
+* Liz Rice has published a set of
+  [basic eBPF examples using libbpfgo](https://github.com/lizrice/libbpfgo-beginners).
+  The eBPF program is trivial, the objective is to get familiar with
+  [libbpfgo](https://github.com/aquasecurity/tracee/tree/main/libbpfgo), a set
+  of bindings in Golang to the libbpf C library.
```
Comment (tklauser) [inline src/posts/2021-02-22-ebpf-updates/index.md:208]:
```suggestion
  of bindings in Go to the libbpf C library.
```

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+
+* Improve XDP performance for the veth by allocating socket buffers in bulks
+  for `ndo_xdp_xmit()`.
+  (Lorenzo Bianconi,
+  [link](https://lore.kernel.org/bpf/a14a30d3c06fff24e13f836c733d80efc0bd6eb5.1611957532.git.lorenzo@kernel.org/t/#u))
+
+## Did You Know? Program size limit
+
+Do you know what is the maximum size of an eBPF program? You may have heard of
+programs limited to 4k instructions, but this is has changed some time ago.
```
Comment (tklauser) [inline src/posts/2021-02-22-ebpf-updates/index.md:375]:
```suggestion
programs limited to 4k instructions, but this has changed some time ago.
```

**Review by tklauser — COMMENTED** [review]
Some spelling fixes inline. LGTM overall. Especially liked the section on program size limits!

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+* [_Networking Performances in the Linux Kernel, Getting the most out of the Hardware_](https://fosdem.org/2021/schedule/event/network_performance_in_kernel/),
+  from Maxime Chevallier.  
+  Definitely centered on networking, this presentation follows the path of
+  packets through the low-level mechanisms involved in the hardware and in the
+  Linux stack. There is not much about eBPF itself, but the last section of the
+  talk helps understand how XDP and AF\_XDP complement the other networking
+  components in the kernel.
+
+* [_Monitoring MariaDB Server with bpftrace on Linux_](https://fosdem.org/2021/schedule/event/mariadb_bpftrace/),
+  This presentation is both an introduction to bpftrace itself and to its
```
[the remark this answers is not in the document]
Comment (qmonnet) [reply src/posts/2021-02-22-ebpf-updates/index.md:64]: Nice catch, thanks! I changed my link at some point, looks like I took away the author when removing the old one :(

File: src/posts/2021-02-22-ebpf-updates/index.md (hunk 1 above)
Comment (qmonnet) [reply src/posts/2021-02-22-ebpf-updates/index.md:146]: :exploding_head:

**Review by tklauser — APPROVED** [review]
:rocket:

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+
+The 4096-instruction limit did not, in fact, disappear entirely. The kernel
+still enforces it for programs loaded by non-root users (more precisely, users
+without `CAP_SYS_ADMIN` or, starting with Linux 5.8, `CAP_BPF`).
+
+eBPF programs tend to be small, and the one-million-state complexity
+limit is big enough that most use cases will never hit it. Some advanced
+projects using eBPF to implement more complex features may be facing it, and
+Cilium for example is
+[regularly adjusting to satisfy to the verifier's requirements](https://github.com/cilium/cilium/issues?q=is%3Aissue+is%3Aopen+complexity+label%3Akind%2Fcomplexity-issue).
```
Comment (pchaigno) [inline src/posts/2021-02-22-ebpf-updates/index.md:423]: Let's maybe not link ot a list of bugs in Cilium :smile:

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+Some ways to work around complexity may include the use of tail calls, of
```
Comment (pchaigno) [inline src/posts/2021-02-22-ebpf-updates/index.md:424]: I'd switched as follows to have consistent constructs in the enumeration:
```suggestion
Some ways to work around complexity may include the use of tail calls and
```

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: src/posts/2021-02-22-ebpf-updates/index.md [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+to termination. Sometimes, some branches “overlap” between several paths, and
+under certain conditions the verifier can skip verifying them after the first
+occurrence. This is called “state pruning”. Without this mechanism, the number
+of instructions to validate would be too high and slow down program loading
+beyond what is acceptable.
+
+When eBPF was introduced, there were two parameters that would limit its size:
+
+- The maximum number of instructions for a program: 4096
+- The complexity limit: 131072 (2<sup>17</sup>)
```
[the remark this answers is not in the document]
Comment (qmonnet) [reply src/posts/2021-02-22-ebpf-updates/index.md:394]: Now you speak of it... Ok I'll double check and update, thanks

File: src/posts/2021-02-22-ebpf-updates/index.md (hunk 7 above)
Comment (qmonnet) [reply src/posts/2021-02-22-ebpf-updates/index.md:423]: Makes sense... Do you have an alternative suggestion or do you think it's better to just drop the link?

Comment (pchaigno) [reply src/posts/2021-02-22-ebpf-updates/index.md:423]: I'd just drop the link.

I think you also need to remove the second `to` ;)

Comment (qmonnet) [reply src/posts/2021-02-22-ebpf-updates/index.md:423]: tssk tssk tssk

**Comment by qmonnet** [comment]
<details><summary>Incremental diff</summary>

``` diff
diff --git a/src/posts/2021-02-22-ebpf-updates/index.md b/src/posts/2021-02-22-ebpf-updates/index.md
index 62df3d36ccf3..c8b195ed3af7 100644
--- a/src/posts/2021-02-22-ebpf-updates/index.md
+++ b/src/posts/2021-02-22-ebpf-updates/index.md
@@ -391,7 +391,7 @@ beyond what is acceptable.
 When eBPF was introduced, there were two parameters that would limit its size:
 
 - The maximum number of instructions for a program: 4096
-- The complexity limit: 131072 (2<sup>17</sup>)
+- The complexity limit: 32768
 
 The second number represents the number of instructions that the verifier is
 allowed to check before forfeiting the verification and rejecting the program.
@@ -401,7 +401,7 @@ So if a program had many logical branches and would require too much effort
 from the verifier, it would fail to load, even if it had fewer than 4096
 instructions.
 
-But both limits were changed in
+But both limits were changed[^limits] in
 [a commit in Linux 5.2](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c04c0d2b968ac45d6ef020316808ef6c82325a82).
 **The complexity limit was raised to one million instructions**. As for the
 maximum limit of instructions, it simply disappeared, meaning that the size of
@@ -419,12 +419,11 @@ without `CAP_SYS_ADMIN` or, starting with Linux 5.8, `CAP_BPF`).
 eBPF programs tend to be small, and the one-million-state complexity
 limit is big enough that most use cases will never hit it. Some advanced
 projects using eBPF to implement more complex features may be facing it, and
-Cilium for example is
-[regularly adjusting to satisfy to the verifier's requirements](https://github.com/cilium/cilium/issues?q=is%3Aissue+is%3Aopen+complexity+label%3Akind%2Fcomplexity-issue).
-Some ways to work around complexity may include the use of tail calls, of
-bounded loops (introduced in Linux 5.3), or reorganizing the code in such a way
-that the number of branches decreases, or that the verifier can prune them more
-efficiently.
+Cilium for example is regularly adjusting to satisfy the verifier's
+requirements. Some ways to work around complexity may include the use of tail
+calls and bounded loops (introduced in Linux 5.3), or reorganizing the code in
+such a way that the number of branches decreases or that the verifier can prune
+them more efficiently.
 
 Hardware offload is yet another story, and has entirely different constraints
 since the program must fit into the hardware's memory. The bound is set as much
@@ -435,6 +434,14 @@ generation and then of the JIT-compiler
 The new, one-million-state complexity limit should be flexible enough for most
 use cases, and in the end, programs have truly one bound: Your imagination!
 
+[^limits]: The complexity limit was actually changed several times since
+[the 32k value from its introduction in Linux 3.18](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=17a5267067f3c372fec9ffb798d6eaba6b5e6a4c):
+[64k in Linux 4.7](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=07016151a446d25397b24588df4ed5cf777a69bb),
+[96k in Linux 4.12](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=3c2ce60bdd3d57051bf85615deec04a694473840),
+[128k in Linux 4.14](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=8e17c1b16277cba0e9426de6fe78817df378f45c),
+and at last
+[1M in Linux 5.2](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c04c0d2b968ac45d6ef020316808ef6c82325a82).
+
 ## Community
 
 The eBPF community keeps growing!
```
</details>

**Pull request merged**